### ISSUE TITLE
Issue while opening multiple sessions in a multithreaded client

### ISSUE THREAD

**Issue by ademyankov** [OP-TEE/optee_os#4478]
**optee_os:** 3.9.x, 3.11, 3.12

The gist is simple. The client app spawns multiple threads; each thread opens a session, invokes a command to TA, and closes the session. optee_os 3.11, 3.12 hangs on the `TEEC_OpenSession` function on one of the threads.

TA is built with `TA_FLAG_SINGLE_INSTANCE | TA_FLAG_MULTI_SESSION | TA_FLAG_INSTANCE_KEEP_ALIVE`. So, it's not concurrent, and all calls to TA are serialized. That was my understanding all the time unless I missed something.

I have some multithreaded code on the REE side that calls to TA. Since I am on an emulator, I define `#define MAX_XLAT_TABLES (16)`, and my multithreaded tests spawn 12 threads. Just for the sake of testing the multithreaded code.

When I moved to OPTEE-3.12, I thought I was chasing a bug (deadlock) in my multithreaded code, but I noticed some OS's bizarre behavior, and eventually, I narrowed it down to the `TEEC_OpenSession` function.

To do a clean and minimalistic test, I modified the hello_world application (host side source code enclosed). It should be easy enough for you guys to use the modified hello app to see it for yourself (don't forget to change MAX_XLAT_TABLES (if spawning more than two threads) and TA flags).

Here are the results of my testing:

**optee_os 3.11, 3.12**
It hangs on `TEEC_OpenSession`. It seems like there is a deadlock in the OS somewhere.

**optee_os 3.9.x**
It works okay-ish since there is no hung, and it basically does what it's supposed to, but OS calls `TA_CreateEntryPoint` **twice**! But I checked it with a static counter; OS destroys the first instance and then opens a second one.

If I create one global session in the main thread, then 3.9 does not call the TA entry point twice, but 3.12 still hangs.

Any thoughts?

Modified hello world:
```
/*
 * Copyright (c) 2016, Linaro Limited
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <err.h>
#include <stdio.h>
#include <string.h>
#include <pthread.h>

/* OP-TEE TEE client API (built by optee_client) */
#include <tee_client_api.h>

/* To the the UUID (found the the TA's h-file(s)) */
#include <hello_world_ta.h>

static TEEC_Context s_ctx;

static void* worker_thread(void* ptr)
{
	TEEC_Result res;
	TEEC_Session sess;
	TEEC_Operation op;
	TEEC_UUID uuid = TA_HELLO_WORLD_UUID;
	uint32_t err_origin;

	/*
	 * Open a session to the "hello world" TA, the TA will print "hello
	 * world!" in the log when the session is created.
	 */
	printf("[%d] TEEC_OpenSession ->\n", *(int*)ptr);
	res = TEEC_OpenSession(&s_ctx, &sess, &uuid,
			       TEEC_LOGIN_PUBLIC, NULL, NULL, &err_origin);
	printf("[%d] <- TEEC_OpenSession\n", *(int*)ptr);

	if (res != TEEC_SUCCESS)
		errx(1, "TEEC_Opensession failed with code 0x%x origin 0x%x",
			res, err_origin);
	/*
	 * Execute a function in the TA by invoking it, in this case
	 * we're incrementing a number.
	 *
	 * The value of command ID part and how the parameters are
	 * interpreted is part of the interface provided by the TA.
	 */
	/* Clear the TEEC_Operation struct */
	memset(&op, 0, sizeof(op));
	/*
	 * Prepare the argument. Pass a value in the first parameter,
	 * the remaining three parameters are unused.
	 */
	op.paramTypes = TEEC_PARAM_TYPES(TEEC_VALUE_INOUT, TEEC_NONE,
					 TEEC_NONE, TEEC_NONE);
	op.params[0].value.a = 42;
	/*
	 * TA_HELLO_WORLD_CMD_INC_VALUE is the actual function in the TA to be
	 * called.
	 */
	printf("[%d] Invoking TA to increment %d\n", *(int*)ptr, op.params[0].value.a);
	res = TEEC_InvokeCommand(&sess, TA_HELLO_WORLD_CMD_INC_VALUE, &op,
				 &err_origin);
	if (res != TEEC_SUCCESS)
		errx(1, "TEEC_InvokeCommand failed with code 0x%x origin 0x%x",
			res, err_origin);
	printf("[%d] TA incremented value to %d\n", *(int*)ptr, op.params[0].value.a);
	/*
	 * We're done with the TA, close the session and
	 * destroy the context.
	 *
	 * The TA will print "Goodbye!" in the log when the
	 * session is closed.
	 */
	printf("[%d] TEEC_CloseSession\n", *(int*)ptr);
	TEEC_CloseSession(&sess);

	return NULL;
}

// To run more threads define:
// #define MAX_XLAT_TABLES (8)
// optee_os_core/arch/arm/plat-vexpress/platform_config.h
// in the section depending on the platform you run the code on
#define WORKERS_COUNT (2)

int main(void)
{
	/* Initialize a context connecting us to the TEE */
	TEEC_Result res = TEEC_InitializeContext(NULL, &s_ctx);
	if (res != TEEC_SUCCESS)
		errx(1, "TEEC_InitializeContext failed with code 0x%x", res);

	pthread_t worker[WORKERS_COUNT];
	int index[WORKERS_COUNT];

	for (int i = 0; i < WORKERS_COUNT; i++) {
		index[i] = i;
		pthread_create(&worker[i], NULL, &worker_thread, &index[i]);
	}

	for (int i = 0; i < WORKERS_COUNT; i++) {
		pthread_join(worker[i], NULL);
	}

	TEEC_FinalizeContext(&s_ctx);
	return 0;
}
```

**Comment by jenswi-linaro** [comment]
It sounds like a problem somewhere. However, I don't have the bandwidth to investigate this at the moment.
It seems it would be useful with this as a proper test case in xtest for this. We'll obviously need a fix to the problem too.
I don't see what `MAX_XLAT_TABLES` has to do with this though.

**Comment by jforissier** [comment]
I don't want to send anybody on a wrong track but I am aware of another issue with single instance TAs and session management: https://github.com/OP-TEE/optee_os/issues/1918, still reproducible today. Might or might not be related... probably not since 1918 involves not closing a session before exiting the CA, but TBH I don't understand very well the mechanisms at play.

**Comment by jenswi-linaro** [comment]
@jforissier, Thanks for the reminder of that bug. We should try to sort out that bug since it probably can show itself in many surprising ways.

**Comment by ademyankov** [comment]
> I don't see what `MAX_XLAT_TABLES` has to do with this though.

Yes, you are right. It has no relationship to a bug itself but rather to the multi-threaded test. If I am not mistaken those tables affect memory mapping, so spawning 4 threads or so consume all the tables (on the armv8 emulator).

**Comment by jenswi-linaro** [comment]
> > I don't see what `MAX_XLAT_TABLES` has to do with this though.
> 
> Yes, you are right. It has no relationship to a bug itself but rather to the multi-threaded test. If I am not mistaken those tables affect memory mapping, so spawning 4 threads or so consume all the tables (on the armv8 emulator).

Are you sure it's needed? The translations tables for TA are allocated from another pool. So if they indeed are needed that might be interesting to look into also.

**Comment by etienne-lms** [comment]
For the record, single instance TA creation deadlock issue #4471 was recently addressed in #4514. I think it corers this issue.
